### PR TITLE
perf: DSV4 chunked prefill (1.7x), attention/router kernel rework, decode I/O overlap

### PR DESCRIPTION
**Pull request by NeelM0906**
## Summary

Performance pass on the DeepSeek V4 Flash path, porting proven techniques from [antirez/ds4](https://github.com/antirez/ds4) and [JustVugg/colibri](https://github.com/JustVugg/colibri). Every change is parity-tested against the existing path; several candidate optimizations were implemented, measured, and deliberately reverted with the rationale recorded.

### Measured (Apple M5, 24 GB, `dsv4.gturbo`, seed 42, 16 slots, identical protocol before/after)

| Metric | Before | After |
|---|---|---|
| Prefill, 629-token prompt | 183.8 s (3.4 tok/s) | **107.0 s (5.9 tok/s), 1.7x** |
| Prefill, 3051-token prompt | ~892 s (extrapolated) | **473.2 s, 1.9x** (first 2051 tok batched, remainder falls back) |
| Decode | 3.60–3.74 tok/s | 3.73–3.81 tok/s (disk-bound; see below) |
| DSV4 context ceiling | ~8.3K hard kernel cap (threadgroup array trap) | no kernel cap (Swift `maxContext` only) |
| Tests | 829 | **857, all green** |

> Long-prompt number updated after review: the original `beae4cd` measured 734.8 s because a span crossing the lightning cutover was rejected wholesale and the whole prompt silently fell back to token-by-token (caught by Codex review). `7e3b789` splits crossing spans so the eligible prefix stays batched — remeasured 473.2 s with identical output text.

Decode output is byte-identical to the pre-change engine in every configuration tested; chunked prefill is bit-exact against the sequential path (full-logits equality, not argmax).

## Changes

**DSV4 chunked prefill** (`RealForwardRunner+DSV4Prefill.swift`, new) — prompts previously ran token-by-token through the decode path. Now layer-major batched chunks: one command buffer per layer per chunk, routed-expert reads grouped per expert and amortized across the chunk (~1.3 TB → ~86 GB read for an 800-token prompt), depth-1 tile pipelining reusing the `PrefillRoutedTileScheduler` discipline. Positions past the lightning-selection cutover (`(indexTopK + 1) × csaRate − 1` = 2051 for the production model) fall back to the sequential path; a span crossing the cutover batches its eligible prefix and replays only the remainder. A mid-chunk failure marks the runner dirty (same `markDirty`/`markCommitted` discipline as `executePrefillChunk`), rejecting further use until `reset()`. `MFERENCE_DSV4_PREFILL=off` escape hatch.

**DSV4 attention decode kernel** — restructured on ds4's heads8 pattern: 8 query heads per threadgroup, each shared K=V row staged once to threadgroup memory and consumed by all 8 heads for both QK and V; online softmax removes the 2176-entry logits array and the context cap it imposed. Parity with the retained reference kernel within 1 fp16 ulp. Grouped o_proj collapsed 8 dispatches → 1 (bit-identical); per-row compressor blit encoders replaced by direct GEMV writes.

**Router top-k on GPU, parallelized** — the sqrtsoftplus k6, hash k6, and k8 selects ran on a single GPU thread; now parallel simdgroup selection (colibri's `r_top8_par` pattern), bit-identical including tie order, proven against the retained serial kernels.

**Decode loop overlap** — an `MTLSharedEvent` signaled immediately after the router top-k write wakes the CPU to plan slots and start expert preads while the same command buffer continues executing the shared-expert FFN (passive wait, no spin — colibri's M-series power-budget finding). `MFERENCE_ROUTER_EVENT=0` reverts to full-buffer waits. `MFERENCE_PHASES` counters now reset at the prefill/decode boundary and split expert I/O into overlapped vs exposed.

**PILOT speculative prefetch** (default **OFF**) — layer L+1's router applied to layer L's post-attention state, encoded in the same command buffer (zero extra syncs), feeding a slot-safe background pread pool with join/eviction-protection/confirmation accounting. Measured 67.5% recall on the real model and −70% exposed decode I/O — but a net wall-clock loss on the 24 GB test machine (speculative read volume saturates the SSD), so it ships disabled. `MFERENCE_SPEC_PREFETCH=pilot` for machines with disk/page-cache headroom.

### Measured-and-reverted (rationale in code comments/tests)
- GEMV activation threadgroup staging: int4 GEMV already at DRAM roofline (143–156 GB/s), int2 gain ~1.5% — below the repo's bar.
- Prefill readahead (`F_RDADVISE` lookahead): the existing 8-way concurrent pread already saturates the disk.
- Temporal-locality prefetch: structurally cannot fire — per-layer slot caches already retain last-token experts (pinned by a test).

## Why decode barely moved

The truthful phase counters show decode on this machine is disk-bound: ~16 s of every 34 s decode window is exposed expert pread, and PILOT proved that hiding it just moves the same bytes elsewhere on a saturated SSD. The kernel/sync work matters on machines where experts fit in page cache (32 GB+), where the GPU path becomes binding. Remaining levers for small-RAM machines: fewer bytes on disk (rANS-style lossless expert compression, ~24%) and more cache slots.

## Test plan
- [x] Full suite: 857 tests / 139 suites green (`Scripts/test.sh`)
- [x] Parity suites: router top-k (bit-identical incl. ties), attention (1 fp16 ulp), o_proj (bit-identical), chunked prefill (full-logits bit-exact across chunk sizes, ragged tails, mixed batched/fallback spans, a cutover-crossing span split mid compressor window, plus a negative control that corrupts the reduce and must fail)
- [x] Determinism: seeded decode output byte-identical across event on/off, spec-prefetch modes, and prefill chunked/sequential
- [x] Real-model benchmarks before/after on identical protocol (numbers above)
- [ ] Community-benchmark protocol run on a larger-RAM machine (M5 Pro/Max) to quantify the kernel-path gains where decode is not disk-bound

### CODE REVIEW

File: Sources/Mference/Runtime/Inference/RealForwardRunner.swift [hunk 1]
```diff
@@ -717,20 +1036,69 @@ public final class RealForwardRunner: ChunkedPrefillRunner, ContextWindowReporti
...
-                done += 1
+            for (spanIndex, span) in spans.enumerated() {
+                let lower = tokens.index(tokens.startIndex, offsetBy: span.tokenOffset)
+                let upper = tokens.index(lower, offsetBy: span.tokenCount)
+                let isLastSpan = spanIndex == spans.count - 1
+                let batched = DSV4ChunkedPrefill.supports(
+                    config: cfg,
+                    startPosition: span.startPosition,
+                    tokenCount: span.tokenCount,
+                    expertCacheSlots: expertSlots)
```
[the remark this answers is not in the document]
Comment (NeelM0906) [reply Sources/Mference/Runtime/Inference/RealForwardRunner.swift]: Good catch — confirmed: with `--prefill-chunk auto` the planner emitted one span for the whole prompt, `supports` rejected it on the final position, and the entire prompt silently replayed token-by-token. This is exactly why our 3051-token benchmark underperformed.

Fixed in 7e3b789: `supports` is now a wrapper over a new `batchedTokenPrefix(config:startPosition:tokenCount:expertCacheSlots:)`, and the dispatch loop batches the eligible prefix of a crossing span and replays only the remainder token-by-token (head emission moves to the remainder's last token). The boundary is `(indexTopK + 1) × csaCompressRate − 1` positions, algebraically identical to the old integer-division check, so no previously-batched span changes behavior.

Verified:
- New test `"span crossing the lightning cutover batches its eligible prefix"`: single 60-token span with the toy's cutover at position 51 — a split mid compressor window — full-logits bit-exact vs sequential decode, including 2 continuation steps.
- Real model (M5, 3051-token prompt, `--prefill-chunk auto`, seed 42): prefill **734.8 s → 473.2 s**, generated text unchanged.
- Full suite: 857 tests green.

File: Sources/Mference/Runtime/Inference/RealForwardRunner.swift [hunk 2]
```diff
@@ -717,20 +1036,69 @@ public final class RealForwardRunner: ChunkedPrefillRunner, ContextWindowReporti
...
+                    let runner: DSV4ChunkedPrefill
+                    if let cached = dsv4Prefill, cached.chunkCapacity >= span.tokenCount {
+                        runner = cached
+                    } else {
+                        runner = try DSV4ChunkedPrefill(
+                            bindings: bindings,
+                            chunkTokens: max(config.chunkTokens, span.tokenCount))
+                        dsv4Prefill = runner
+                    }
+                    let greedy = try await runner.run(tokens: tokens[lower..<upper],
```
[the remark this answers is not in the document]
Comment (NeelM0906) [reply Sources/Mference/Runtime/Inference/RealForwardRunner.swift:1093]: Agreed — the batched runner advances DSV4 layer state (window rings, compressor pending state, indexer caches) as it encodes, so a throw from an expert `pread` or a Metal command-buffer error mid-chunk left that state ahead of the KV cursor while `requireClean` still passed.

Fixed in 7e3b789: the batched span call is now bracketed exactly like `executePrefillChunk` — `prefillChunkState.markDirty(startPosition:tokenCount:)` immediately before `runner.run`, `markCommitted()` only after the KV cursor advances. A mid-chunk failure now makes any subsequent `produce`/`prefillChunked` throw `chunkedRunnerDirty` until `reset()`, matching the existing chunked path's contract. Full suite (857 tests) green; the token-by-token fallback path keeps its pre-existing semantics (it advances per token, same as decode).
